Add Node debugging guide

paul-vd opened the pull request:
Adds a guide for node debug in front-commerce-skeleton using vscode or chrome devtools

Preview: https://deploy-preview-377--heuristic-almeida-1a1f35.netlify.app/docs/reference/debugging.html

### Debugging with VSCode
https://user-images.githubusercontent.com/39598117/162025005-36a22fb8-f6e5-4c7f-b3e4-16f27336c5bd.mp4



### Debugging with Chrome DevTools
https://user-images.githubusercontent.com/39598117/162024726-6d247230-9d36-4b4f-a1c2-c6de433a4922.mp4

dpobel commented on source/docs/advanced/features/debugging.md:
```diff
@@ -0,0 +1,98 @@
+Once the server starts, open a new tab in Chrome and visit `chrome://inspect`, where you should see your Front-Commerce application inside the **Remote Target** section. Click **inspect** under your application to open a separate DevTools window, then go to the **Sources** tab.
+
+<!--
+TODO: Add development source maps with webpack?
```
what do you mean by that ? source map is already enabled normally (even if there
are several flavors).

paul-vd replied:
True

paul-vd commented on source/docs/advanced/features/debugging.md:
```diff
@@ -0,0 +1,98 @@
+
+### Client-side code
+
+Start your development server as usual by running `front-commerce start`, `npm run start`. Once the server starts, open `http://localhost:4000` (or your alternate URL) in Chrome. Next, open Chrome's Developer Tools (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>J</kbd> on Windows/Linux, <kbd>⌥</kbd>+<kbd>⌘</kbd>+<kbd>I</kbd> on macOS), then go to the **Sources** tab.
```
> also it would be nice to not be 100% Chrome oriented and to mention that it also works well for client-side with Firefox 
I think this is related to what we have discussed a retro, we support a set of conditions and environments and anything added to that is additional, this is only as a point of reference to debug, based on the majority use case, although not limited to chrome, using chrome as reference supports a wider base of browsers/users (chromium).

And because for the server-side code we also supply docs for chrome dev tools node debugger, it makes sense to include it for the front end as well.

paul-vd replied:
I did add some reference to other clients and also inspector clients with links to docs.

paul-vd commented on source/docs/advanced/features/debugging.md:
```diff
@@ -0,0 +1,98 @@
+To learn more about how to use a JavaScript debugger, take a look at the following documentation:
+
+- [Node.js debugging in VS Code: Breakpoints](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_breakpoints)
+- [Chrome DevTools: Debug JavaScript](https://developers.google.com/web/tools/chrome-devtools/javascript)
```
Firefox has not been included because it does not support node.js debugging which was the main point of this documentation.

https://stackoverflow.com/a/56759070

dpobel replied:
yeah I know but it works perfectly for the Frontend and unless we want a webkit only world, I think it should be mentioned here.

paul-vd replied:
Add reference to firefox as well

paul-vd commented on source/docs/advanced/features/debugging.md:
```diff
@@ -0,0 +1,98 @@
+
+### Server-side code
+
+To debug server-side Front-Commerce code with Chrome DevTools, you need to pass the [`--inspect`](https://nodejs.org/api/cli.html#cli_inspect_host_port) flag to the underlying Node.js process:
```
The docs are specifically aimed at either VS code debugger or chrome. dev tools (nodejs debugger), I will add a video example to show the diff between just a browser-based debugger (chrome, firefox, etc..) and nodejs debugger inside chrome.

paul-vd replied:
I added an example for both VSCode & Chrome DevTools

dpobel replied:
I know Firefox does not support that but as mentioned in https://nodejs.org/en/docs/guides/debugging-getting-started/#inspector-clients there are others debuggers than Chrome dev tools (VSCode is one of them btw).

dpobel replied:
I think you misunderstand my point. My suggestion above was just a little rephrasing so that the sentence is a bit more open. The way it is written seems to suggest that only Chrome Dev Tools can be used which is wrong. Any debugger implementing the correct protocol can be used.

paul-vd replied:
> also some months ago, I added a paragaph about debugging with the debugger in https://developers.front-commerce.com/docs/reference/environment-variables.html#Debugging so it would be nice to unify that with what you add. But besides this and the remarks 👍

Hmm, should I maybe move this debugging to that documentation?

paul-vd replied:
Added back links between the two docs

Kagaminara commented on source/docs/advanced/features/debugging.md:
````diff
@@ -0,0 +1,88 @@
+}
+```
+
+If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `3000` in `http://localhost:3000` with the port you're using instead, you can use the the url defined in your `FRONT_COMMERCE_URL`
````
Is it a typo?
```suggestion
If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `4000` in `http://localhost:4000` with the port you're using instead, you can use the the url defined in your `FRONT_COMMERCE_URL`
```

paul-vd replied:
Hehe yea typo, been years that I work with `3000` as the default 😬

dpobel commented:
and don't forget:
![image](https://user-images.githubusercontent.com/305563/162688065-62cbf890-f2bc-4aed-94ed-bd487394b571.png)

paul-vd commented:
> and don't forget: ![image](https://user-images.githubusercontent.com/305563/162688065-62cbf890-f2bc-4aed-94ed-bd487394b571.png)

I updated the routing so that the debugging falls under `References` because `Features` was technically the wrong placement for this doc.

I also updated the blockquote mentioned in your image to backlink to the Debugging docs, and visa versa in the debugging docs.

abailleulfc commented on source/docs/reference/debugging.md:
````diff
@@ -0,0 +1,93 @@
+}
+```
+
+If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `4000` in `http://localhost:4000` with the port you're using instead, you can use the the url defined in your `FRONT_COMMERCE_URL`
````
```suggestion
If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `4000` in `http://localhost:4000` with the port you're using instead, you can use the url defined in your `FRONT_COMMERCE_URL`
```

paul-vd replied:
lol, I have no idea why I always do this, `the the` 🤦‍♂️

abailleulfc commented on source/docs/reference/debugging.md:
```diff
@@ -0,0 +1,93 @@
+
+If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `4000` in `http://localhost:4000` with the port you're using instead, you can use the the url defined in your `FRONT_COMMERCE_URL`
+
+Now in vscode go to the Debug panel (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> on Windows/Linux, <kbd>⇧</kbd>+<kbd>⌘</kbd>+<kbd>D</kbd> on macOS), select a launch configuration, then press <kbd>F5</kbd> or select **Debug: Start Debugging** from the Command Palette to start your debugging session.
```
```suggestion
Now in VS Code go to the Debug panel (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> on Windows/Linux, <kbd>⇧</kbd>+<kbd>⌘</kbd>+<kbd>D</kbd> on macOS), select a launch configuration, then press <kbd>F5</kbd> or select **Debug: Start Debugging** from the Command Palette to start your debugging session.
```

real34 commented on source/docs/reference/debugging.md:
````diff
@@ -0,0 +1,93 @@
+}
+```
+
+If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `4000` in `http://localhost:4000` with the port you're using instead, you can use the url defined in your `FRONT_COMMERCE_URL`
````
```suggestion
If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `4000` in `http://localhost:4000` with the port you're using instead. You can use the url defined in your `FRONT_COMMERCE_URL`.
```

real34 commented on source/docs/reference/debugging.md:
```diff
@@ -0,0 +1,93 @@
+
+If you're [changing the port number](/docs/reference/environment-variables.html#Host) your application starts on, replace the `4000` in `http://localhost:4000` with the port you're using instead, you can use the url defined in your `FRONT_COMMERCE_URL`
+
+Now in VS Code go to the Debug panel (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> on Windows/Linux, <kbd>⇧</kbd>+<kbd>⌘</kbd>+<kbd>D</kbd> on macOS), select a launch configuration, then press <kbd>F5</kbd> or select **Debug: Start Debugging** from the Command Palette to start your debugging session.
```
For advanced settings or understand how it works in detail, I think it makes sense to link to the official documentation here (as it will be where developers will stop when reading the documentation).

```suggestion
Now in VS Code go to the Debug panel (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> on Windows/Linux, <kbd>⇧</kbd>+<kbd>⌘</kbd>+<kbd>D</kbd> on macOS), select a launch configuration, then press <kbd>F5</kbd> or select **Debug: Start Debugging** from the Command Palette to start your debugging session. See [Debugging in Visual Studio Code (official documentation)](https://code.visualstudio.com/docs/editor/debugging) for more details.
```

real34 commented on source/docs/reference/debugging.md:
```diff
@@ -0,0 +1,93 @@
+- [Chrome DevTools: Debug JavaScript](https://developers.google.com/web/tools/chrome-devtools/javascript) (Server & Client)
+- [The Firefox JavaScript Debugger](https://firefox-source-docs.mozilla.org/devtools-user/debugger/index.html) (Client Only)
+- [Node.js Inspector Clients](https://nodejs.org/en/docs/guides/debugging-getting-started/#inspector-clients)
+- [Debug with log output](http://localhost:4444/docs/reference/environment-variables.html#Debugging)
```
```suggestion
- [Debug with log output](/docs/reference/environment-variables.html#Debugging)
```

real34 commented on source/docs/reference/debugging.md:
```diff
@@ -0,0 +1,93 @@
+
+## Debugging with Chrome DevTools
+
+Most modern browsers support Client Side Debugging, which allows you to debug your Front-Commerce application in the browser. But if you would like to debug Server-side code in a browser that does not support Node.js debugging, you can refer to the [Inspect Clients](https://nodejs.org/en/docs/guides/debugging-getting-started/#inspector-clients) docs to find a debugging client which suites your en ...
```
Let's be friendly with Firefox users :joy: :gun: 
```suggestion
Several browsers support Client Side Debugging, which allows you to debug your Front-Commerce application in the browser. If your browser does not support Node.js debugging, you can refer to the [Inspect Clients](https://nodejs.org/en/docs/guides/debugging-getting-started/#inspector-clients) docs to find a server-side code debugging client which suits your environment.
```

real34 reviewed (approved):
Thanks for this doc!

Since you've recorded videos, don't you think we could try to embed them in each section to start experimenting with video content too?

paul-vd commented:
> Thanks for this doc!
> 
> Since you've recorded videos, don't you think we could try to embed them in each section to start experimenting with video content too?

Good idea!
